### ISSUE TITLE
Records detached when converting nested multiset

### ISSUE THREAD

**Issue by dillon-giacoppo** [jOOQ/jOOQ#12269]
When nesting multisets, the inner records become detached during convert. Even when the inner type is a record it will always will create a new `CTX.configuration()` due to https://github.com/jOOQ/jOOQ/blob/main/jOOQ/src/main/java/org/jooq/impl/MultisetDataType.java#L145.

I am not sure if this is expected behaviour or if I am doing something incorrectly (which may well be the case 😅).

If this is unexpected behaviour I think the fix should just be to inspect the type of the object during conversion and passthrough the configuration if its a Record.

### Background

I have a DB with relation `... - N -> Variant - N -> VariantFlag - 1 -> Flag`. I have tried converting the retrieval of this relation to nested multisets. The data structure I am mapping this to basically just groups the records together:

`{ VariantRecord, List< { VariantFlagRecord, FlagRecord } > }` but when done only the `VariantRecord` is attached.

The jOOQ code to setup and convert the multiset is:

```java
private static Field<List<VariantRecordBundle>> getVariantSelector(DSLContext ctx) {
    var flags =
        DSL.multiset(
                ctx.select()
                    .from(VARIANT_FLAG)
                    .join(FLAG)
                    .on(FLAG._ID.eq(VARIANT_FLAG.FLAG_ID))
                    .where(VARIANT_FLAG.VARIANT_ID.eq(VARIANT._ID)))
            .convertFrom(
                rl ->
                    rl.map(
                        r ->
                            new VariantFlagRecordBundle(
                                r.into(VARIANT_FLAG), r.into(FLAG))));  // These records are detached.

    return DSL.multiset(
            ctx.select(VARIANT.fields())
                .select(flags)
                .from(VARIANT)
                .where(VARIANT.VERSION_ID.eq(VERSION._ID)))
        .convertFrom(
            rl ->
                rl.map(
                    r ->
                        new VariantRecordBundle(
                            r.into(VARIANT), r.into(flags).value1())));  // VARIANT record is attached in the outer query.
}
```

### Versions

jOOQ: org.jooq:jooq:3.15.1
Java: 16
Database (include vendor): 8.0.22 MySQL Community Server - GPL (docker)
OS: MacOS
JDBC Driver (include name if inofficial driver): mysql:mysql-connector-java:8.0.24

**Comment by lukaseder** [comment]
Thanks a lot for your report. All records should be attached if they're fetched from the database, including nested records, so this is likely a bug.

Will look into this soon.

**Comment by lukaseder** [comment]
I'm trying to reproduce this, but am failing. All nested `Result` and `Record` instances are attached. Can you try again, showing a more minimal example? We have a template to create MCVEs, if this helps: https://github.com/jOOQ/jOOQ-mcve

**Comment by lukaseder** [comment]
Just stumbled upon this in code:

```java
result.add(newRecord(true, recordType, row, CONFIG)
```

... and remembered this issue. The `CONFIG` is the internal default configuration, which is the wrong one. I'll investigate this soon, with this in mind, I can probably reproduce the problem.

**Comment by lukaseder** [comment]
Related issue, the records have their `changed` flags set to `true`: https://github.com/jOOQ/jOOQ/issues/13403

**Comment by lukaseder** [comment]
Yep, can reproduce it with the same integration test as I wrote for #13403. Both problems are caused by the same `MultisetDataType::convert` and `RecordDataType::convert` methods, which should be no-ops if the argument is already of the correct type.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13406), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13407)
